crates/storage/src/connection/state_update.rs

```diff
@@ -475,6 +480,237 @@ pub(super) fn contract_class_hash(
     .map_err(|e| e.into())
 }
 
+pub(super) fn reverse_updates(
+    tx: &Transaction<'_>,
+    from: BlockNumber,
+    to: BlockNumber,
+) -> anyhow::Result<HashMap<ContractAddress, ReverseContractUpdate>> {
+    let storage_updates = reverse_storage_updates(tx, from, to)?;
+    let nonce_updates: Vec<(ContractAddress, Option<ContractNonce>)> =
+        reverse_nonce_updates(tx, from, to)?;
+    let contract_updates = reverse_contract_updates(tx, from, to)?;
+
+    let mut updates: HashMap<ContractAddress, ReverseContractUpdate> = Default::default();
+
+    for (contract_address, class_hash_update) in contract_updates {
+        updates
+            .entry(contract_address)
+            .or_insert_with(|| match class_hash_update {
+                None => ReverseContractUpdate::Deleted,
+                Some(_) => ReverseContractUpdate::Updated(ReverseContractUpdateDetails {
+                    class: class_hash_update,
+                    ..Default::default()
+                }),
+            });
+    }
+
+    for (contract_address, nonce_update) in nonce_updates {
+        if let Some(update) = updates
+            .entry(contract_address)
+            .or_insert_with(|| ReverseContractUpdate::Updated(Default::default()))
+            .update_mut()
+        {
+            update.nonce = nonce_update
+        };
+    }
+
+    for (contract_address, storage_updates) in storage_updates {
+        if let Some(update) = updates
+            .entry(contract_address)
+            .or_insert_with(|| ReverseContractUpdate::Updated(Default::default()))
+            .update_mut()
+        {
+            update.storage = storage_updates
+        };
+    }
+
+    Ok(updates)
+}
+
+type StorageUpdates = Vec<(StorageAddress, Option<StorageValue>)>;
+
+fn reverse_storage_updates(
+    tx: &Transaction<'_>,
+    from_block: BlockNumber,
+    to_block: BlockNumber,
+) -> anyhow::Result<HashMap<ContractAddress, StorageUpdates>> {
+    let mut stmt = tx.inner().prepare(
+        r"WITH
+        updated_addresses(contract_address, storage_address) AS (
+            SELECT DISTINCT
+                contract_address, storage_address
+            FROM storage_updates
+            WHERE
+                block_number > ?2 AND block_number <= ?1
```

Comment:
I initially got confused with the naming and assumed `from > to`, which here is the other way round because of descending order of blocks when reverting.